core/client/fs/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -145,39 +152,30 @@ public BlockInStream getInStream(long blockId, Protocol.OpenUfsBlockOptions open
           Preconditions.checkNotNull(options.getUfsReadLocationPolicy(),
               PreconditionMessage.UFS_READ_LOCATION_POLICY_UNSPECIFIED);
       address = blockLocationPolicy
-          .getWorker(GetWorkerOptions.defaults().setBlockWorkerInfos(getWorkerInfoList())
+          .getWorker(GetWorkerOptions.defaults().setBlockWorkerInfos(getEligibleWorkers())
               .setBlockId(blockId).setBlockSize(blockInfo.getLength()));
-      if (address == null) {
-        throw new UnavailableException(ExceptionMessage.NO_WORKER_AVAILABLE.getMessage());
-      }
     } else {
       // TODO(calvin): Get location via a policy.
-      // Although blockInfo.locations are sorted by tier, we prefer reading from the local worker.
-      // But when there is no local worker or there are no local blocks, we prefer the first
-      // location in blockInfo.locations that is nearest to memory tier.
-      // Assuming if there is no local worker, there are no local blocks in blockInfo.locations.
-      // TODO(cc): Check mContext.hasLocalWorker before finding for a local block when the TODO
-      // for hasLocalWorker is fixed.
-      for (BlockLocation location : blockInfo.getLocations()) {
-        WorkerNetAddress workerNetAddress = location.getWorkerAddress();
-        if (workerNetAddress.getHost().equals(mLocalHostName)) {
-          address = workerNetAddress;
+      List<TieredIdentity> locations = blockInfo.getLocations().stream()
+          .map(location -> location.getWorkerAddress().getTieredIdentity())
+          .collect(Collectors.toList());
+      Optional<TieredIdentity> nearest = mTieredIdentity.nearest(locations);
+      if (nearest.isPresent()) {
+        address = blockInfo.getLocations().stream()
+            .map(BlockLocation::getWorkerAddress)
+            .filter(a -> a.getTieredIdentity() == nearest.get())
+            .findFirst().get();
+        if (mTieredIdentity.getTier(0).getTierName().equals(Constants.LOCALITY_NODE)
+            && mTieredIdentity.topTiersMatch(nearest.get())) {
           source = BlockInStreamSource.LOCAL;
-          break;
-        }
-      }
-      if (address == null) {
-        // No local worker/block, choose a random location. In the future we could change this to
-        // only randomize among locations in the highest tier, or have the master randomize the
-        // order.
-        List<BlockLocation> locations = blockInfo.getLocations();
-        if (locations.isEmpty()) {
-          throw new UnavailableException(ExceptionMessage.NO_WORKER_AVAILABLE.getMessage());
+        } else {
+          source = BlockInStreamSource.REMOTE;
         }
-        address = locations.get(mRandom.nextInt(locations.size())).getWorkerAddress();
```

Comment:
Where is the non-"local" randomness built in?

Reply:
good catch, added a `Collections.shuffle` before calling `nearest`, and added units tests to cover this method